### DISCUSSION TITLE
Export Garmin GPX

### DISCUSSION THREAD

**Discussion by WolfgangZ42** [devemux86/cruiser#501] [Q&A] [answered]
First of all, I'm really glad I don't have to use a Garmin device. In contrast, planning and navigating with the Cruiser app is pure pleasure! 

But: there are friends who can't get rid of Garmin and I would like to make tours available to them from time to time. With kurviger.de directly or by a detour via gpxshaping.de this is possible. Kurviger Pro (K1) can also do it (via the server).

How would it be if we also enabled a special Garmin export with Cruiser? So in addition to GPX, JSON and Kurviger, we would also offer a **Garmin Shaping Point GPX**. Another advantage for Cruiser!

**Comment by devemux86** [comment]
@WolfgangZ42 thanks for the suggestion.

Can you post a route in normal GPX and Garmin GPX to compare the formats?

What are the differences?

**Reply by WolfgangZ42** [reply]
@devemux86 sure you can get an example. I took the "Notzingen"-Tour ([Notzingen.-.Wangen.im.Allgau.kurviger.txt](https://github.com/devemux86/cruiser/files/15050883/Notzingen.-.Wangen.im.Allgau.kurviger.txt)), modified the original Cruiser-Export ( [Notzingen - Wangen im Allgäu.gpx.txt](https://github.com/devemux86/cruiser/files/15050888/Notzingen.-.Wangen.im.Allgau.gpx.txt)) and tested that ([Notzingen - Wangen im Allgäu_mod.gpx.txt](https://github.com/devemux86/cruiser/files/15050890/Notzingen.-.Wangen.im.Allgau_mod.gpx.txt)) with BaseCamp and Navigator VI. 

The differences are shown here with compare in Notepad++:
<img width="800" src="https://github.com/devemux86/cruiser/assets/121806238/62d2ec93-579c-4a87-932d-4820c3b69d2f"/>

Would be great to get a Garmin version directly from Cruiser export.

**Reply by devemux86** [reply]
@WolfgangZ42 thanks for marking the differences, this is what I needed. :+1: 

- Is it only about the `rte`? The `wpt` and `trk` remain the same?

- We just need to add for via points and start / destination `<extensions><trp:ViaPoint/></extensions>`
and for shaping points `<extensions><trp:ShapingPoint/></extensions>` ?

- What is calculation mode? Doesn't it work without it?

- Is something needed at `rte`? Why transportation mode?

I want to keep things simple with the least possible Garmin additions.

**Reply by 0709wiwiwi** [reply]
> What is calculation mode? Doesn't it work without it?

Thats what I found.
[Route Calculation Mode](https://www8.garmin.com/manuals/webhelp/drive51-61/EN-US/GUID-B9E024D7-8317-4401-983C-9D6B6FA08AC6.html)

> Is something needed at rte? Why transportation mode?   

Untraceable without Zumo.  Zumo users ?  Possibly the Zumo also has a car profile mode ?
One difference between car and motorbiking is that cars do not route through narrower streets (less than 3 m) ? 

I think to add the profile preferences is important.
 As the Garmins do recalculate after (rte) route imports the profile should imo be fairly correct.
This to prevent that a Motorcycle desing does not run over nearby foot or cycle paths for example.

See image Basecamp.
![Basecamp profielen](https://github.com/devemux86/cruiser/assets/54167201/0bddf71f-5342-492a-bf0b-5fde7e9fe523)

**Reply by devemux86** [reply]
@0709wiwiwi these seem to be needed for calculations (like in Cruiser).

What these have to do with storage formats that only have to store coordinates?

**Reply by 0709wiwiwi** [reply]
As far I understand Garmin after import has to recalculate the route this to attach the turninstructions.

The difference.

- An **ORIGINAL** Garmin (Basecamp) gpx (rte) route export adds the simple track (without turnpoints) into the route points (complex) extension.

- The simple gpx navigation track adds the turninstructions in trackpoints  and additionally the (snapped to road) shaping or via routepoints into the trackpoints type tag.

**Reply by WolfgangZ42** [reply]
> What is calculation mode? Doesn't it work without it?

> Is something needed at rte? Why transportation mode?

I have to admit that I simply took "calculation mode" and "transportation mode" from GPXshaping, because I assume that this was checked accordingly. Perhaps "FasterTime" is the default - I haven't found anything about it. Not even about "transportation mode". I will now do a test without this information.

**Reply by devemux86** [reply]
@WolfgangZ42 

> I will now do a test without this information.

Please test it and let's keep the minimum needed extra tags, so it is easier to maintain.

**Reply by WolfgangZ42** [reply]
Creating .gpx files w/o calculation, transportation and both ... all could be imported into BaseCamp and sent to the Navigator VI. And all of them looks alike:
![2024-04-21 12 25 00](https://github.com/devemux86/cruiser/assets/121806238/0e316b84-af76-4564-8a89-58eed8ba9672)
Don't ask me why the totaldistance is 131 km instead of 153 km ... in all 4 cases.
Unfortunately I can't drive the route: I've no bike with BMW-cradle, the weather is bad and I'm not in Notzingen. 🥴😎

**Reply by WolfgangZ42** [reply]
This is the .gpx file with minimum definition: [Notzingen - Wangen im Allgäu_mob.gpx.txt](https://github.com/devemux86/cruiser/files/15051558/Notzingen.-.Wangen.im.Allgau_mob.gpx.txt) - I think we could start with this format.

**Reply by devemux86** [reply]
@WolfgangZ42 thanks for the tests!

What about the `<sym>`? Are they really needed?

**Reply by WolfgangZ42** [reply]
as an extension to the above comparison:

<img width="800" src="https://github.com/devemux86/cruiser/assets/121806238/a3c86cae-36a8-4bd5-8eca-1e9906ff798f"/>

**Reply by WolfgangZ42** [reply]
> What about the <sym>? Are they really needed?

Don't know - but in GPXshaping output there is no 'sym' and it works. So it may be omitted for Garmin GPX - and only stay in the "normal" .gpx file.

**Reply by devemux86** [reply]
A final question: the example files have `rte` and `trk`.
What about  the `wpt`, should they have anything special?

**Reply by WolfgangZ42** [reply]
I have never used wpt but will have a look at GPXshaping and come back to you later. My grandchildren are keeping me busy this afternoon.

**Reply by 0709wiwiwi** [reply]
> What about the <sym>? Are they really needed?

Any sym or no sym ok. Without sym or sym not supported by the gps (app) a default icon is displayed.

Minimal (Garmin) rte to test.  (route_routepoint coordinates_yes, timestamps_no, name_no, sym_no, type_no, extension_yes).
[Notzingen.-.Wangen.im.Allgau_mob_minimal.gpx.txt](https://github.com/devemux86/cruiser/files/15052087/Notzingen.-.Wangen.im.Allgau_mob_minimal.gpx.txt)

> What about the wpt, should they have anything special?

Special ? no, by gpx wpt you do indicate individual locations.  Share a (custom) poi location(s) for example. 
[Expressions in gpx](https://github.com/devemux86/cruiser/discussions/250#discussion-5425220l)

**Reply by 0709wiwiwi** [reply]
> Don't ask me why the totaldistance is 131 km instead of 153 km 

It is the (straight) route line distance. Or bird track distance ;-). The GPS has to construct the track to be driven first. Only after that calculation the correct drive distance can be determined. 
Even with very many routepoints, some smaller deviations are still possible. Only the kurviger or the gpx navtrack file guarantees 100% accuracy after an import.

**Reply by devemux86** [reply]
> Only the kurviger or the gpx navtrack file guarantees 100% accuracy after an import.

And the JSON format, which is the origin of the Kurviger format.

**Reply by 0709wiwiwi** [reply]
> First of all, I'm really glad I don't have to use a Garmin device. In contrast, planning and navigating with the Cruiser app is pure pleasure!

Definitely an understandable reaction. Even experts who have been developing tools specifically for Garmin for almost a quarter of a century don't seem to have unravelled all complex variant (extensions) methods used.

[Routeprofiles](https://www.javawa.nl/routeprofielen.html)

**Comment by devemux86** [comment]
We are not talking to support every possible Garmin extension, like colors, etc.

Just the basics so a route can be opened in a Garmin device with all waypoints.

cc @0709wiwiwi

**Reply by 0709wiwiwi** [reply]
UPDATE: The above description by Wolfgang as indicated should be sufficient. 

_Also this.  "Demo to show that the instructions in a gpx navtrack are discretely attached"
Even the export as gpx rte route including Garmin extension along with the gpx navigation track format (instead of the naked gpx track) should also be possible.   Or the whole package all at once ;-)_
Notice. The exported rtept's are "snapped" to the road network.

All at once next.
[Notzingen - Wangen im Allgäu_rte_garmin_navtrk.gpx.txt](https://github.com/devemux86/cruiser/files/15050993/Notzingen.-.Wangen.im.Allgau_rte_garmin_navtrk.gpx.txt)

**Reply by DIrgi** [reply]
This is a Tour exported from the Garmin Basecamp Software (including Waypoints and Shaping Points):
[PT2302 Mogadouro-Torre S_d_Estrella 282km RT.zip](https://github.com/devemux86/cruiser/files/15049130/PT2302.Mogadouro-Torre.S_d_Estrella.282km.RT.zip)

This is the same Tour developed in Kurviger, exported as kurviger navigation file:
[02 Mogadouro-Torre S.d.Estrella 282km.kurviger.zip](https://github.com/devemux86/cruiser/files/15049139/02.Mogadouro-Torre.S.d.Estrella.282km.kurviger.zip)

This is the same file developed in kurviger, exported as Garmin file with shaping points:
[02 Mogadouro-Torre S.d.Estrella 282km.gpx.zip](https://github.com/devemux86/cruiser/files/15049141/02.Mogadouro-Torre.S.d.Estrella.282km.gpx.zip)

**Comment by devemux86** [comment]
@WolfgangZ42 and everyone,

An export with 1 via and 1 shaping from Cruiser with the Garmin GPX format: [garmin.gpx.txt](https://github.com/devemux86/cruiser/files/15051707/garmin.gpx.txt)

Please check it if it is correct and if we need anything else.

**Reply by 0709wiwiwi** [reply]
@WolfgangZ42 
To test with the next file.  (Contains multiple consecutive routepoints to test by Garmin gps) 
[Notzingen.-.Wangen.im.Allgau_mob_minimal.gpx.txt](https://github.com/devemux86/cruiser/files/15052138/Notzingen.-.Wangen.im.Allgau_mob_minimal.gpx.txt)

Anyway I am not sure if rte minimalistic is advisable notice next
https://github.com/devemux86/cruiser/discussions/501#discussioncomment-9177845

**Comment by devemux86** [comment]
<img src="https://github.com/devemux86/cruiser/assets/3484020/d158c52e-d82d-405a-8fe5-a07380156544" width="50%"/>

**Comment by MotoUKRider** [comment]
I've mentioned a couple of times how Cruiser loads a Garmin BaseCamp GPX just nicely (observing the Shaping+Via-Points perfectly).  So, how do things work in reverse?  First, I looked at the file using ITNConv which found both a "Route" and "Track".  Safari-so-goody...

Here's the initial import into BaseCamp:
![garmin-route-1](https://github.com/devemux86/cruiser/assets/125696109/93924cc0-7bb0-4cf6-b351-96199035e69a)

Here's what it looks like after a route calculation - this is entirely based on my routing profile, so a GPS unit will probably so the same thing if you use the "Route" info rather than the "Track".
![garmin-route-2](https://github.com/devemux86/cruiser/assets/125696109/1ced7f59-529f-4bec-bc29-9fd9dfcb1494)

Also, my ported map theme looks nice in BaseCamp and refreshes a possibly-soon-to-be-forgotten bit of mapping software :grinning:

**Reply by WolfgangZ42** [reply]
Both ("garmin" and "Notzingen.-.Wangen.im.Allgau_mob_minimal") could be imported into BaseCamp and transmitted to Navigator VI w/o any problem.

**Reply by devemux86** [reply]
@WolfgangZ42 thanks for the tests!

So it seems we are ok with the new export Garmin option.

**Comment by devemux86** [comment, marked as answer]
Export Garmin GPX is available in [Cruiser 4.0.0](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

**Comment by MotoUKRider** [comment]
Be warned that if you create a route in Cruiser app and then export it as a "Waypoint, Garmin" GPX, any existing Favorites/Waypoints in BaseCamp will be overwritten with a basic icon (and remove the elevation).  "Route, Garmin" is the selection to make if you want to get a route into your desktop system without making a mess :smile: 

Here's the offending GPX - I lost the camera+fuel-pump icons on these Favorites.  I can see why this happened (the Favorite fields in Cruiser are limited compared to BaseCamp), but wanted to put something here before somebody loaded a really *big* route into it!

`<?xml version='1.0' encoding='UTF-8' standalone='no' ?>
<gpx xmlns="http://www.topografix.com/GPX/1/1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.topografix.com/GPX/1/1 http://www.topografix.com/GPX/1/1/gpx.xsd" version="1.1" creator="Cruiser" xmlns:trp="http://www.garmin.com/xmlschemas/TripExtensions/v1">
  <metadata>
    <name>20240504_101439WG</name>
  </metadata>
  <wpt lat="54.1191005490147" lon="-2.4540502034649148" />
  <wpt lat="54.146523009985685" lon="-2.469614865258336">
    <name>Fuel, Ingleton</name>
  </wpt>
  <wpt lat="54.20727252960205" lon="-2.362639904022217">
    <name>Photo, Ribblehead Viaduct</name>
  </wpt>
  <wpt lat="54.303712" lon="-2.1932380000000062" />
  <wpt lat="54.305252" lon="-2.098234999999995" />
</gpx>`

**Reply by devemux86** [reply]
@MotoUKRider do you mean we need an improvement somewhere?

Garmin was added for the route exports (not favorites) and in the `rtept`.

**Reply by MotoUKRider** [reply]
@devemux86 It was more of a warning, as I was working out how to get the routes *from* Cruiser onto my desktop (not a common thing for me).

As I understand it, the format of the file dictates how BaseCamp "handles" the file - so when it finds a GPX with `wpt` entries, it uses its internal process to either import them as new items or manipulate an existing item.  The latter happened here!

This actually works to your advantage if you get a plain GPX/KML of places and want to assign a nice icon to them: Export them from BaseCamp, then open the file and perform a Find/Replace on the `sym` value to match your desired icon, then re-import and it'll just update the existing entries with that icon.

What I hadn't realised until now is that with a *very* basic WPT GPX, it effectively scrubs the existing contents in BaseCamp.

**Comment by 0709wiwiwi** [comment]
@MotoUKRider 

Consider what follows still a "game" with no extra value ;-)

You didn't explain how you arrived at your Cruiser production.
You may have selected Cruiser favourites as routepoints ?
In your example Cruiser Petrol & Photopoint icon selections. 
Similar icons as triggered by Garmin sym Gas Station and Scenic Area.

**Should Cruiser apply a best matching Garmin sym in the gpx export of the favourites (icons) your setup might succeed ;-)**

Your example.

- Modified as gpx waypoints with best matching Garmin sym and gpx type (no problem for Garmin).
- Then converted these waypoints to a gpx route_routepoints with Garmin via_shapingextension as an extra .

From the gpx route then produced a Cruiser_Locus gpx navigation track.

[MotoUK.zip](https://github.com/devemux86/cruiser/files/15209984/MotoUK.zip)

![Garmin Mapsource import all in file wpt rte navtrk](https://github.com/devemux86/cruiser/assets/54167201/25a1f1d1-5097-40cd-82ec-ae9789c7ddf2)

[Garmin_imi_all_in.gpx.txt](https://github.com/devemux86/cruiser/files/15211059/Garmin_imi_all_in.gpx.txt)
